ClassLoader issue using plugin aggregate function in window accumulator: java.lang.NoClassDefFoundError

I'm seeing exceptions caused by ClassLoaders when using a plugin aggregator function plugin inside a window function. The issue is similar to the issue mentioned on [this Slack thread](https://trinodb.slack.com/archives/CP1MUNEUX/p1640390600083200?thread_ts=1640349840.080000&cid=CP1MUNEUX).

Specifically, during this call in AccumulatorCompiler it is using the ClassLoader for the AccumulatorCompiler class, not the ClassLoader for the aggregation function state. This works for aggregation functions within Trino, but those that are plugins do not work.
`Class<? extends WindowAccumulator> windowAccumulatorClass = defineClass(definition, WindowAccumulator.class, callSiteBinder.getBindings(), classLoader);`

The exception I am getting is this:
```
com.google.common.util.concurrent.ExecutionError: java.lang.NoClassDefFoundError: io/gefisher/trino/operator/aggregation/state/TestAggregatorState
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2053)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3966)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4863)
        at com.google.common.cache.ForwardingCache.get(ForwardingCache.java:55)
        at io.trino.collect.cache.CacheUtils.uncheckedCacheGet(CacheUtils.java:28)
...
Caused by: java.lang.NoClassDefFoundError: io/gefisher/trino/operator/aggregation/state/TestAggregatorState
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:398)
        at com.google.common.reflect.Reflection.initialize(Reflection.java:64)
        at io.airlift.bytecode.ClassGenerator.defineClasses(ClassGenerator.java:177)
        at io.airlift.bytecode.ClassGenerator.defineClass(ClassGenerator.java:117)
        at io.trino.util.CompilerUtils.defineClass(CompilerUtils.java:63)
        at io.trino.util.CompilerUtils.defineClass(CompilerUtils.java:57)
        at io.trino.operator.aggregation.AccumulatorCompiler.generateWindowAccumulatorClass(AccumulatorCompiler.java:287)
        at io.trino.operator.window.AggregationWindowFunctionSupplier.<init>(AggregationWindowFunctionSupplier.java:40)
        at io.trino.sql.planner.LocalExecutionPlanner$Visitor.lambda$getWindowFunctionImplementation$5(LocalExecutionPlanner.java:1194)
```

Do we need a similar fix to this? https://github.com/trinodb/trino/pull/10408

I propose we do something like this in `AccumulatorCompiler.generateWindowAccumulatorClass(...)`. This works:
```java
List<ClassLoader> classLoaders = metadata.getAccumulatorStateDescriptors().stream()
                .map(stateDescriptor-> stateDescriptor.getStateInterface().getClassLoader())
                .distinct()
                .collect(Collectors.toList());
if (classLoaders.size() != 1) {
    throw new TrinoException(StandardErrorCode.FUNCTION_IMPLEMENTATION_ERROR, "Window accumulator function state interface does not have a single class loader.");
}

DynamicClassLoader classLoader = new DynamicClassLoader(classLoaders.stream().findFirst().get(), AccumulatorCompiler.class.getClassLoader());
```

* Is this an issue? Or is there a workaround?
* Can I send out a PR to fix this?
* Is there a preference for how to fix this more cleanly than my suggestion above?

--- a/core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java
+++ b/core/trino-main/src/main/java/io/trino/operator/aggregation/AccumulatorCompiler.java
@@ -219,6 +219,9 @@ public final class AccumulatorCompiler
             AggregationImplementation implementation,
             FunctionNullability functionNullability)
     {
+        // change types used in Aggregation methods to types used in the core Trino engine to simplify code generation
+        implementation = normalizeAggregationMethods(implementation);
+
         DynamicClassLoader classLoader = new DynamicClassLoader(AccumulatorCompiler.class.getClassLoader());
 
         List<Boolean> argumentNullable = functionNullability.getArgumentNullable()

--- a/core/trino-main/src/test/java/io/trino/operator/aggregation/TestAccumulatorCompiler.java
+++ b/core/trino-main/src/test/java/io/trino/operator/aggregation/TestAccumulatorCompiler.java
@@ -15,11 +15,16 @@ package io.trino.operator.aggregation;
 
 import com.google.common.collect.ImmutableList;
 import io.airlift.bytecode.DynamicClassLoader;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
 import io.trino.operator.aggregation.state.StateCompiler;
+import io.trino.operator.window.InternalWindowIndex;
 import io.trino.server.PluginManager;
 import io.trino.spi.Page;
 import io.trino.spi.block.Block;
 import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.block.Int96ArrayBlock;
+import io.trino.spi.block.LongArrayBlockBuilder;
 import io.trino.spi.function.AccumulatorState;
 import io.trino.spi.function.AccumulatorStateFactory;
 import io.trino.spi.function.AccumulatorStateSerializer;
@@ -33,6 +38,8 @@ import io.trino.sql.gen.IsolatedClass;
 import org.testng.annotations.Test;
 
 import java.lang.invoke.MethodHandle;
+import java.lang.reflect.Constructor;
+import java.util.Optional;
 
 import static io.trino.operator.aggregation.AggregationFunctionAdapter.AggregationParameterKind.INPUT_CHANNEL;
 import static io.trino.operator.aggregation.AggregationFunctionAdapter.AggregationParameterKind.STATE;
@@ -96,7 +103,20 @@ public class TestAccumulatorCompiler
         // test if we can compile aggregation
         AccumulatorFactory accumulatorFactory = AccumulatorCompiler.generateAccumulatorFactory(signature, implementation, functionNullability);
         assertThat(accumulatorFactory).isNotNull();
-        assertThat(AccumulatorCompiler.generateWindowAccumulatorClass(signature, implementation, functionNullability)).isNotNull();
+
+        // compile window aggregation
+        Constructor<? extends WindowAccumulator> actual = AccumulatorCompiler.generateWindowAccumulatorClass(signature, implementation, functionNullability);
+        assertThat(actual).isNotNull();
+        WindowAccumulator windowAccumulator;
+        try {
+            windowAccumulator = actual.newInstance(ImmutableList.of());
+        }
+        catch (ReflectiveOperationException e) {
+            throw new RuntimeException(e);
+        }
+        // call the functions to ensure that the code does not reference the wrong state
+        windowAccumulator.addInput(new TestWindowIndex(), 0, 5);
+        windowAccumulator.evaluateFinal(new LongArrayBlockBuilder(null, 1));
 
         TestingAggregationFunction aggregationFunction = new TestingAggregationFunction(
                 ImmutableList.of(TIMESTAMP_PICOS),
@@ -119,5 +139,75 @@ public class TestAccumulatorCompiler
             TIMESTAMP_PICOS.writeObject(builder, new LongTimestamp(i, i));
         }
         return builder.build();
+    }
+
+    private static class TestWindowIndex
+            implements InternalWindowIndex
+    {
+        @Override
+        public int size()
+        {
+            return 10;
+        }
+
+        @Override
+        public boolean isNull(int channel, int position)
+        {
+            return false;
+        }
+
+        @Override
+        public boolean getBoolean(int channel, int position)
+        {
+            return false;
+        }
+
+        @Override
+        public long getLong(int channel, int position)
+        {
+            return 0;
+        }
+
+        @Override
+        public double getDouble(int channel, int position)
+        {
+            return 0;
+        }
+
+        @Override
+        public Slice getSlice(int channel, int position)
+        {
+            return Slices.EMPTY_SLICE;
+        }
+
+        @Override
+        public Block getSingleValueBlock(int channel, int position)
+        {
+            return null;
+        }
+
+        @Override
+        public Object getObject(int channel, int position)
+        {
+            return null;
+        }
+
+        @Override
+        public void appendTo(int channel, int position, BlockBuilder output)
+        {
+            output.appendNull();
+        }
+
+        @Override
+        public Block getRawBlock(int channel, int position)
+        {
+            return new Int96ArrayBlock(1, Optional.empty(), new long[] {0}, new int[] {0});
+        }
+
+        @Override
+        public int getRawBlockPosition(int position)
+        {
+            return 0;
+        }
     }
 }

